sdk/monitor/Azure.Monitor.Ingestion/src/LogsIngestionClient.cs

```diff
@@ -196,6 +212,142 @@ public virtual Response<UploadLogsResult> Upload<T>(string ruleId, string stream
+        /// <remarks> See error response code and error response message for more detail. </remarks>
+#pragma warning disable AZC0004 // DO provide both asynchronous and synchronous variants for all service methods.
+        public virtual async Task<Response<UploadLogsResult>> UploadAsync<T>(string ruleId, string streamName, IEnumerable<T> logEntries, UploadLogsOptions options, CancellationToken cancellationToken = default)
+#pragma warning restore AZC0004 // DO provide both asynchronous and synchronous variants for all service methods.
+        {
+            Argument.AssertNotNullOrEmpty(ruleId, nameof(ruleId));
+            Argument.AssertNotNullOrEmpty(streamName, nameof(streamName));
+            Argument.AssertNotNullOrEmpty(logEntries, nameof(logEntries));
+
+            // Calculate the number of threads to use.
+            // If there are 0 workers, method will run serially. Otherwise will run in parallel with number of workers given.
+            int _maxWorkerCount = options.MaxConcurrency == 0 ? DefaultWorkerCount : options.MaxConcurrency;
+            using var scope = ClientDiagnostics.CreateScope("LogsIngestionClient.Upload");
+
+            RequestContext requestContext = GenerateRequestContext(cancellationToken);
+            Response response = null;
+            List<UploadLogsError> errors = new List<UploadLogsError>();
+
+            try
+            {
+                scope.Start();
+                // A list of tasks that are currently executing which will
+                // always be smaller than MaxWorkerCount
+                List<Task<RunningTask<T>>> runningTasks = new();
+                // Partition the stream into individual blocks
+                foreach (BatchedLogs<T> batch in Batch(logEntries))
+                {
+                    // Start staging the next batch (but don't await the Task!)
+                    Task<RunningTask<T>> task = CommitBatchListAsync(
+                        batch,
+                        ruleId,
+                        streamName,
+                        cancellationToken);
+
+                    // Add the block to our task and commit lists
+                    runningTasks.Add(task);
+
+                    // If we run out of workers
+                    if (runningTasks.Count >= _maxWorkerCount)
+                    {
+                        // Wait for at least one of them to finish
+                        Task<RunningTask<T>> finished = await Task.WhenAny(runningTasks).ConfigureAwait(false);
+                        // Clear any completed blocks from the task list
+                        for (int i = 0; i < runningTasks.Count; i++)
+                        {
+                            Task runningTask = runningTasks[i];
+                            if (!runningTask.IsCompleted)
+                            {
+                                continue;
+                            }
+
+                            await runningTask.ConfigureAwait(false);
```

Comment:
We shouldn't need to await this task as we already know that it is completed.

Reply:
We chatted about this - the desired behavior is that we don't want to throw if the server returned a response error. In this case we are not throwing anywhere so the task will not actually throw. If we get an unexpected exception, we do want to throw, which should happen as part of the `await Task.WhenAll` call. This line isn't necessary as the task is completed, and we will already handle propagating exceptions on L315.